### PR TITLE
Adding ACT with residual refinement

### PR DESCRIPTION
**Pull request by jmarangola**
This pull request implements ACT behavioral cloning policies which can be trained as a base policy and further refined with residual PPO. 

#### Behavioral Cloning
A state-based Action Chunking Transformer (ACT) policy was trained with minimal hyperparameter tuning, achieving a success rate of 44.5% on the `one_leg` task with low randomness. The corresponding tensorboard can be accessed [here](https://wandb.ai/johnmarangola-/one_leg-state-low/runs/ese28gy3?nw=nwuserjohnmarangola).

Behavioral Cloning Performance

**Hyperparameters for the state-based ACT policy**:  
- Max Learning Rate: 1e-5  
- Batch Size: 256  
- Prediction Horizon: 32  
- Action Horizon: 1  
- Observation Horizon: 1  
- Dimensions of latent space: 32
- Transformer FFN Dimensions: 3200
- KL Divergence Loss Weight: 0.1  
- Reconstruction Loss: L1

All other parameters were set to the defaults specified in the respective base training/actor configurations.

#### Residual Refinement
This BC policy was refined using PPO, with default parameters as outlined in the original paper and base configurations. This refinement yielded a >96% success rate under the same conditions. The results can be viewed in the tensorboard [here](https://wandb.ai/johnmarangola-/one_leg-residual-rl/runs/yucvpuzk?nw=nwuserjohnmarangola).
Screenshot 2024-10-18 at 11 34 30 PM

### Implementation Details

The ACT was trained as a state-based conditional variational autoencoder (CVAE) with a transformer encoder as the encoder and a transformer encoder/decoder architecture as the (CVAE) decoder. ADAM was used as the optimizer for the base policy with the hyperparameters specified below. 

_Note_: To be precise here, the words "observation" and "state" are used interchangeably below _although_ I am  to is always referring to the (normalized) "state" of the environment, either the object poses or the proprioceptive state of the robot (or both in some cases).

The CVAE encoder, represented as `p(z | s_t, a_t, ..., a_{t + Tp})`, is a transformer-based module distinct from the primary transformer used in the CVAE decoder. It takes as input the current robot state embedding, along with a "chunk" of ground truth actions.

The encoder learns the posterior (outputs a latent representation `z`, encapsulating (`mu`, `logvar`)). A zero-mean, unit-variance Gaussian is used as the prior, and the analytical solution for the KL divergence between the learned distribution and the prior is incorporated into the loss by using the reparameterization trick to allow for backpropagation through the sampling process. The aforementioned CVAE encoder is implemented as a transformer encoder with multiple transformer encoder layers. Each transformer encoder layer uses multi-head (self) attention and global positional encodings to encode token position locations globally. Following self-attention in each transformer encoder layer, we have the standard norm and add with the output of a small, 2 layer FFN, plus dropout throughout for regularization. Transformer layers are applied sequentially, using the output of the previous layer as the input to the next, starting with the tokenized input sequence.

The (CVAE) decoder is implemented as a transformer encoder/decoder architecture where the transformer encoder is the same as the transformer encoder mentioned above. The CVAE decoder takes as input a latent representation `z` and  normalized observation embeddings for proprioceptive state and part poses and outputs a chunk (sequence) of continuous, normalized actions `(a_t, ..., a_{t + Tp})`. 

The transformer encoder in the CVAE decoder is similar to the transformer encoder in the CVAE encoder, but it has 6 transformer encoder layers instead of 4. Each encoder layer performs self-attention output from the previously layer using the query embeddings. the self-attention uses the embeddings formed from robot state and latent representations as the initial input (for k=0). The transformer leverages additional positional _embeddings_ to learn modal differences between representations (namely latent, priorioceptive and object state) which are added to these respective embeddings (this approach works well given the markovian observation assumption).

The transformer decoder consists of layers that perform multi-head self-attention on the input sequence and the query embeddings, followed by multi-head cross-attention using the "memory" from the encoder. The input sequence at each iteration `k+1` comes from the output of the `k'th` iteration with k=0 being all zeros. The transformer decoder outputs a sequence of hidden states, which are projected back to the action space using an "action head" to get a chunk of actions. Under the hood, the transformer decoder is made up of transformer decoder layers that have the same auxiliary components as the transformer encoder layers (FFN, dropout, layer norm). In addition to the attention mechanisms, the transformer decoder also stacks intermediate outputs from each layer into a tensor and returns, instead of returning the final output from the last layer.

The final component of the loss is the reconstruction loss computed between the predicted chunk of actions and the actions inputted to the CVAE encoder (from expert demonstrations). An L1 loss was used for this reconstruction loss term. The full loss function minimized can be expressed as: 

<p align="center">
  equation (3)
</p>

Where R corresponds to the reconstruction loss and \beta_{KL} corresponds to the KLD weight. 

During inference, we throw away the CVAE encoder and only use the decoder, sampling from the prior instead of the posterior to obtain `z`.

#### Assumptions

1. Markov Assumption: Since this implementation was a quick MVP built in a few days, it was suggested to start simple with a behavior cloning (BC) policy under the Markov Assumption. Consequently, the observation horizon is currently limited to a single timestep. Experiment (1) generalizes this and I have a branch for training state-based ACT policies which take a sequence of state observations instead of a single observation. 
2. No support for variable length sequences: Since the training and inference pipeline is already well-implemented and provides/requires fixed-length sequences, there is no need to handle variable-length sequences. As a result, implementing causal attention mechanisms with padding, padding heads, or causal masking was unnecessary, as observations during training and inference are already constrained by default to the pre-defined observation horizon in the config. In principle, this is quite simple to implement, but adds clutter to the implementation if not required, so it was omitted.

### Experiments
1. ***Non-markovian observation extension:*** I experimented with adding a branch that generalizes the ACT to allow for sequences of temporal observations to be encoporated as input to the policy (`obs_horizon > 1`). Doing so required some relatively straightforward changes such as adding positional encodings to states (in time) and slightly modifying the way that modal positional embeddings are applied in the CVAE decoder. As part of this experiment, I also added a set of learned modal embeddings for the proprioceptive state and the actions for the transformer encoder (CVAE encoder). While this approach appears promising, I did not get a chance to finish a quantitative evaluation and it is likely that there could be other issues that make this approach perform worse such as causal confusion, irrelevant history, etc. In practice, this might not be the worst assumption, although it is nice to make things general :)
2. ***Experimentation with hyperparameters of base policy:*** Several hyperparmeters were experimented with very briefly including the learning rate, feedforward dimension for the transformer encoder and transformer decoder, size of the latent space, and the action and prediction horizon as well as the KL divergence weights. Details have been omitted here for brevity but in general it was found that decreasing the max learning rate and action horizon and increasing the prediction horizon had a relatively significant effect on policy performance. There are several runs which contain results of various experiments. 
3. ***torchscript/JIT compilation support:*** I have a branch that torchscripts the model floating around but did not have time to test it. Sometimes one must settle for an MVP, unfortunately.

### CODE REVIEW

File: src/behavior/base.py [hunk 1]
```diff
@@ -686,3 +688,132 @@ def set_task(self, task):
         Set the task for the actor
         """
         pass
+
+
+class RefinementActor(nn.Module):
```
Comment (jmarangola) [inline src/behavior/base.py]: This was an experimental attempt to unify residual refinement by making a singular shared class which takes as input a particular actor into the constructor. This ended up being more complex than simply overriding the Actor in the base class (as was the previous implementation pattern) so it is not used by the final solution here but I left it in this PR because I believe the idea could lead to a more modular approach that would automatically generalize to any Actor added.

File: src/behavior/act.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+        self.positional_embeddings = nn.Embedding(3, self.dim_hidden)
+
+    def forward(self, robot_state: Tensor, obj_state: Tensor,
+                actions: Optional[Tensor] = None
+                ) -> Tuple[Tensor, Tensor, Tuple[Tensor, Tensor]]:
+        """
+        Forward pass for the ACT as a CVAE.
+
+        Args:
+            robot_state (Tensor): Robot state embedding (B, obs_horizon, R)
```
Comment (jmarangola) [inline src/behavior/act.py:355]: Todo: this doctrine needs to be updated. These shapes are incorrect.
